### PR TITLE
ISSUE 364 - Set up TMail event bus, event for new contacts, listener to index

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/364

### CODE REVIEW

File: tmail-backend/events/src/main/scala/com/linagora/tmail/events/TmailContactListener.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+class TmailContactListener @Inject()(contactSearchEngine: EmailAddressContactSearchEngine) extends ReactiveGroupEventListener {
+  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[TmailContactListener])
+
+  override def getDefaultGroup: Group = PushListenerGroup()
+
+  override def reactiveEvent(event: Event): Publisher[Void] = {
+    event match {
+      // TODO, EmailAddressContactSearchEngine.index will be updated
+      case contactEvent: TmailContactUserAddedEvent => handlerContactUserAddedEvent(contactEvent)
```
Comment (vttranlina) [inline tmail-backend/events/src/main/scala/com/linagora/tmail/events/TmailContactListener.scala:20]: // TODO,
Need to rebase after https://github.com/linagora/tmail-backend/issues/362 already

File: tmail-backend/events/src/main/scala/com/linagora/tmail/events/TmailEvent.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+trait TmailEvent extends Event {
+
+}
+
+trait TmailContactUserEvent extends TmailEvent {
+
+}
+
+case class TmailContactUserAddedEvent(eventId: EventId, username: Username, contact: MailAddress, firstname: String, lastname: String) extends TmailContactUserEvent {
```
Comment (chibenwa) [inline tmail-backend/events/src/main/scala/com/linagora/tmail/events/TmailEvent.scala:15]: IMO we should reuse the `ContactFields` POJO from https://github.com/linagora/tmail-backend/pull/384

File: tmail-backend/events/src/main/scala/com/linagora/tmail/events/TmailEvent.scala [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.apache.james.core.{MailAddress, Username}
+import org.apache.james.events.Event
+import org.apache.james.events.Event.EventId
+
+trait TmailEvent extends Event {
+
+}
+
+trait TmailContactUserEvent extends TmailEvent {
```
Comment (chibenwa) [inline tmail-backend/events/src/main/scala/com/linagora/tmail/events/TmailEvent.scala:11]: IMO everything related to contacts should be moved close to the contact API

**Comment by vttranlina** [comment]
IMO not need a Publisher class. We can use directly `eventBus.dispatch`

**Review by chibenwa — APPROVED** [review]
Fair. We will extract this when it makes sense.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala [hunk 4]
```diff
@@ -35,7 +38,20 @@ trait EmailAddressContactSearchEngine {
   def index(domain: Domain, fields: ContactFields): Publisher[EmailAddressContact]
 
   def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact]
+}
+
+trait TmailEvent extends Event {
+}
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala:44]: I think you can remove the brackets?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala [hunk 5]
```diff
@@ -35,7 +38,20 @@ trait EmailAddressContactSearchEngine {
...
+
+trait TmailContactUserEvent extends TmailEvent {
+}
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala:47]: Idem

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContactListenerIntegrationTest.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    eventBus.register(new EmailAddressContactListener(searchEngine))
+  }
+
+  @Test
+  def shouldIndexContactWhenHasContactUserAddedEvent(): Unit = {
+    eventBus.dispatch(CONTACT_ADDED_EVENT, NO_KEYS).block()
+    assertThat(SFlux.fromPublisher(searchEngine.autoComplete(ACCOUNT_ID, "contact1"))
+      .map(_.fields.address)
+      .collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(MAIL_ADDRESS)
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContactListenerIntegrationTest.scala:43]: Shouldn't we check we have the name too?

File: tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContactListenerIntegrationTest.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Test
+  def indexShouldBeIdempotent(): Unit = {
+    eventBus.dispatch(CONTACT_ADDED_EVENT, NO_KEYS).block()
+    eventBus.dispatch(CONTACT_ADDED_EVENT, NO_KEYS).block()
+    assertThat(SFlux.fromPublisher(searchEngine.autoComplete(ACCOUNT_ID, "contact1"))
+      .map(_.fields.address)
+      .collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(MAIL_ADDRESS)
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContactListenerIntegrationTest.scala:53]: Idem

**Review by Arsnael — COMMENTED** [review]
And don't forget to rebase on master :)

**Review by quantranhong1999 — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala (hunk 4 above)
Comment (vttranlina) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala:44]: yeah, wonder syntax

**Review by Arsnael — APPROVED** [review]

**Pull request merged**